File: backend/app/schemas/audit.py
from datetime import datetime

from pydantic import BaseModel, Field


class AuditEventRead(BaseModel):
    id: str
    actor_id: str | None
    action_type: str
    entity_type: str
    entity_id: str | None
    occurred_at: datetime
    request_id: str | None
    before_snapshot: dict | None
    after_snapshot: dict | None
    metadata_: dict | None = Field(alias="metadata_", default=None)

    model_config = {"from_attributes": True}


class AuditListResponse(BaseModel):
    items: list[AuditEventRead]
    total: int
    limit: int
    offset: int

File: frontend/src/api/audit.ts
import { apiFetch } from "./client";

export interface AuditEventRead {
  id: string;
  actor_id: string | null;
  actor_display?: string | null;
  action_type: string;
  entity_type: string;
  entity_id: string | null;
  entity_display?: string | null;
  occurred_at: string;
  request_id?: string | null;
  before_snapshot: Record<string, unknown> | null;
  after_snapshot: Record<string, unknown> | null;
  metadata_: Record<string, unknown> | null;
  detail_summary?: string | null;
}

export interface AuditListResponse {
  items: AuditEventRead[];
  total: number;
  limit: number;
  offset: number;
}

export interface AuditFilters {
  actor_id?: string;
  action_type?: string;
  entity_type?: string;
  entity_id?: string;
  limit?: number;
  offset?: number;
}

export const auditApi = {
  list: (filters: AuditFilters = {}) => {
    const params = new URLSearchParams();
    if (filters.actor_id) params.set("actor_id", filters.actor_id);
    if (filters.action_type) params.set("action_type", filters.action_type);
    if (filters.entity_type) params.set("entity_type", filters.entity_type);
    if (filters.entity_id) params.set("entity_id", filters.entity_id);
    params.set("limit", String(filters.limit ?? 50));
    params.set("offset", String(filters.offset ?? 0));
    return apiFetch<AuditListResponse>(`/audit?${params.toString()}`);
  },
};
